settings: Add a new realm-level setting, `moderation_request_channel`.

PieterCK opened the pull request:
This PR adds a new realm-level setting `moderation_request_channel`. This realm setting will be used for the future "flag for abuse" function (#20047). 


Fixes part of #20047.
[CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/report.20message.20feature.20.28.2320047.29)
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
`/api/get-events`
![image](https://github.com/user-attachments/assets/db45a239-750b-43a3-a374-01c7d4cd2db2)

`/api/register-queue`
![image](https://github.com/user-attachments/assets/a430c183-9f3d-4474-969e-085205051763)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/models/realms.py:
```diff
@@ -392,6 +392,13 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+    )
```
This isn't a good place to put this in the block; between two different types of announcements channels. Before `new_stream_announcements_stream` seems OK.

timabbott commented on zerver/models/realms.py:
```diff
@@ -1211,6 +1223,7 @@ def get_realm_with_settings(realm_id: int) -> Realm:
         "new_stream_announcements_stream",
         "signup_announcements_stream",
         "zulip_update_announcements_stream",
+        "abuse_report_channel",
```
Keep new fields alphabetical when adding them to an already alphabetized list like this. It's super valuable for readability.

PieterCK replied:
Does this, to some degree, apply to how we'd like to organize the functions too? More specifically, I mean grouping them by category (e.g., realm channel settings functions) and, within each "category", arranging them in alphabetical order by function name. (In addition to other rules, like placing helper functions before the main function they support.)

timabbott commented on zerver/migrations/0639_realm_abuse_report_channel.py:
```diff
@@ -0,0 +1,36 @@
+) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    Realm.objects.exclude(new_stream_announcements_stream__isnull=True).update(
+        abuse_report_channel=F("new_stream_announcements_stream")
```
Did we specify a recommended default for existing organizations? I'm kinda guessing we want this to default to `None`, since abuse reports generally will want to go to a private channel.

PieterCK replied:
Hmm, yeah, I wanted to ask about this because the setup for `signup_announcement_stream` looks similar (no initial value), but in its migration file, it’s given the same initial value, `new_stream_announcements_stream`.

PieterCK replied:
```python
# 0121_realm_signup_notifications_stream

def set_initial_value_for_signup_notifications_stream(
    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
) -> None:
    realm_model = apps.get_model("zerver", "Realm")
    realms = realm_model.objects.exclude(notifications_stream__isnull=True)
    for realm in realms:
        realm.signup_notifications_stream = realm.notifications_stream
        realm.save(update_fields=["signup_notifications_stream"])
```

timabbott commented:
Posted some feedback. I think beyond that, you can probably squash the commits -- conceptually they're all doing one thing, which is adding a non-operational abuse report channel field to the server/API. 

I think you're also missing necessary hunks in the data import/export code paths. Please do a careful `git grep` audit of the existing stream-valued settings on realm objects and make sure you're touching all the files that they touch.

PieterCK commented:
- Squashed commits.
- Added the patch for `import_realm`.
  ```diff
  diff --git a/zerver/lib/import_realm.py b/zerver/lib/import_realm.py
  index 09492f913a..334d7b7446 100644
  --- a/zerver/lib/import_realm.py
  +++ b/zerver/lib/import_realm.py
  @@ -1203,6 +1203,7 @@ def do_import_realm(import_dir: Path, subdomain: str, processes: int = 1) -> Rea
           update_model_ids(PresenceSequence, data, "presencesequence")
   
       # Now we prepare to import the Realm table
  +    re_map_foreign_keys(data, "zerver_realm", "abuse_report_channel", related_table="stream")
       re_map_foreign_keys(
           data, "zerver_realm", "new_stream_announcements_stream", related_table="stream"
       )
  ```
- Rearranged new functions, documentation & parameters.
- Removed `set_initial_value_for_abuse_report_channel` from the migration file.

PieterCK commented on version.py:
```diff
@@ -34,7 +34,7 @@
 # entries in the endpoint's documentation in `zulip.yaml`.
 
-API_FEATURE_LEVEL = 325  # Last bumped for can_administer_channel_group
+API_FEATURE_LEVEL = 327  # Last bumped for realm-level setting, `abuse_report_channel`.
```
level 326 is https://github.com/zulip/zulip/pull/32573

PieterCK commented:
- Renamed `abuse_report_channel` -> `moderation_request_channel`.
- Added an assertion to make sure only private channels are allowed for a Realm's `moderation_request_channel` field
  <details>
  
  <summary>  git diff  </summary>
  
  ```diff
  diff --git a/zerver/actions/realm_settings.py b/zerver/actions/realm_settings.py
  index c439cbe9d6..4a2878b1ff 100644
  --- a/zerver/actions/realm_settings.py
  +++ b/zerver/actions/realm_settings.py
  @@ -434,6 +434,8 @@ def do_set_realm_stream(
   def do_set_realm_moderation_request_channel(
       realm: Realm, stream: Stream | None, stream_id: int, *, acting_user: UserProfile | None
   ) -> None:
  +    if stream is not None and stream.is_public():
  +        raise AssertionError("Only private channels are offered for this setting.")
       do_set_realm_stream(
           realm, "moderation_request_channel", stream, stream_id, acting_user=acting_user
       )
  ```
  </details>

timabbott commented:
Let me know when you have this passing tests; I'm imagining we can roughly integrate this checkpoint of database schema + API support for setting the value, once it's ready, and be free of future merge conflicts/renumbering related to them.

PieterCK commented:
- Renamed `abuse_report_channel` → `moderation_request_channel`.
- Added an assertion to make sure only private channels are allowed when setting a Realm's `moderation_request_channel` field
  <details>
  
  <summary>  git diff  </summary>
  
  ```diff
  diff --git a/zerver/actions/realm_settings.py b/zerver/actions/realm_settings.py
  index c439cbe9d6..4a2878b1ff 100644
  --- a/zerver/actions/realm_settings.py
  +++ b/zerver/actions/realm_settings.py
  @@ -434,6 +434,8 @@ def do_set_realm_stream(
   def do_set_realm_moderation_request_channel(
       realm: Realm, stream: Stream | None, stream_id: int, *, acting_user: UserProfile | None
   ) -> None:
  +    if stream is not None and stream.is_public():
  +        raise JsonableError(_("Only private channels are offered for this setting."))
       do_set_realm_stream(
           realm, "moderation_request_channel", stream, stream_id, acting_user=acting_user
       )
  ```
  </details>
- Added [channels: Prevent moderation_request_channel to be public.](https://github.com/zulip/zulip/pull/32566/commits/14dd0d504ae9806a2b260bf8b22ab5ed81e43dec)

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -426,6 +431,16 @@ def do_set_realm_stream(
+    realm: Realm, stream: Stream | None, stream_id: int, *, acting_user: UserProfile | None
+) -> None:
+    if stream is not None and stream.is_public():
+        raise JsonableError(_("Only private channels are offered for this setting."))
```
"Moderation request channel must be private." seems a clearer and simpler for this? I want to avoid the "offered for this setting" framing, and I'm not sure we'll end up reusing the string for anything else.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4841,6 +4841,19 @@ paths:
+                                      description: |
+                                        The ID of the private channel to which messages flagged for moderation
+                                        requests by users are sent. Moderators can use this channel to review and
+                                        act on reported content.
```
"flagged by users for moderation" might be clearer wording.

PieterCK replied:
Updated to:
```yaml

  The ID of the private channel to which messages flagged by users for
  moderation are sent. Moderators can use this channel to review and
  act on reported content.
```

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4841,6 +4841,19 @@ paths:
+                                        Will be `-1` if moderation requests are disabled.
+
+                                        Since these automated messages are sent by the server, this field is
+                                        primarily relevant to clients containing UI for changing it.
```
I think this is wrong -- it should probably say "Clients should use whether moderation requests are disabled to determine whether to present a report message feature in their UI within a given organization."

PieterCK replied:
hmm did you mean "Clients should **check** whether moderation requests are disabled to ..." or am I misunderstanding the message?

PieterCK replied:
I've updated this for now,:
```
Clients should check whether moderation requests are disabled to
determine whether to present a report message feature in their UI
within a given organization.
```
Let me know if this matches with what you had in mind!

PieterCK commented:
Updated the PR, outstanding issue:
- Question on this comment: https://github.com/zulip/zulip/pull/32566#discussion_r1890316723

timabbott commented on zerver/views/streams.py:
```diff
@@ -324,6 +324,10 @@ def update_stream_backend(
 
+    # Ensure that a moderation request channel isn't set to public.
+    if not proposed_is_private and user_profile.realm.get_moderation_request_channel() == stream:
+        raise JsonableError(_("A moderation request channel cannot be public."))
```
Changed this to use the same string that you had elsewhere. Every string is extra work for translators, so only add new ones if they serve a useful purpose to users.

timabbott commented:
Merged, thanks @PieterCK! Excited to have this chunk of preparation for the feature fully integrated.
